Dynamically shaped image support for `tf.image.resize_image_with_crop_or_pad`

gaohuazuo opened the pull request:
This patch is intended for issue https://github.com/tensorflow/tensorflow/issues/521

A new keyword argument `dynamic_shape=False` is added to these functions in `tensorflow.python.image_ops`:
`crop_to_bounding_box` `pad_to_bounding_box` `resize_image_with_crop_or_pad` `_ImageDimensions`

The change is fully backward compatible. The new functionality of `resize_image_with_crop_or_pad` is tested and worked well. However, the dynamic version of `crop_to_bounding_box` `pad_to_bounding_box` is missing many argument checks compared to the static version.

Keeping both a dynamic and a static version of the same function is quite ugly. Maybe we should keep only the dynamic version?

**TODO**
Test cases

gaohuazuo committed:
Dynamic shape support improvement

Allow `resize_image_with_crop_or_pad` to work with dynamically shaped image.

gaohuazuo committed:
Fix docs issues

tensorflow-jenkins commented:
Can one of the admins verify this patch?

martinwicke commented on tensorflow/python/ops/image_ops.py:
```diff
@@ -478,40 +492,46 @@ def crop_to_bounding_box(image, offset_height, offset_width, target_height,
   image = ops.convert_to_tensor(image, name='image')
-  _Check3DImage(image, require_static=True)
-  height, width, _ = _ImageDimensions(image)
+  _Check3DImage(image, require_static=(not dynamic_shape))
```
These not... are ugly. Can you change the argument from dynamic_shape=False to static_shape_check=True? Then all the functions have the same kind of parameter.

martinwicke commented:
This is great. Can you add tests for this? I think having the ability to use static checks is useful (sometimes), and backwards compatibility is nice, so I don't think you should remove he static checks.

gaohuazuo commented:
Should these functions simply skip checks that cannot be performed at graph construction time? That would eliminate the need  of a new argument.

martinwicke commented:
That would mean that tensors of unknown size will pass all the static tests? That's a good idea. I think the danger is minimal here. The errors in the resize kernels are not too bad anyway.

gaohuazuo committed:
Removed `dynamic_shape` argument

gaohuazuo commented on tensorflow/python/ops/image_ops.py:
```diff
@@ -427,44 +433,57 @@ def pad_to_bounding_box(image, offset_height, offset_width, target_height,
   after_padding_height = target_height - offset_height - height
 
-  if after_padding_width < 0:
+  if not isinstance(offset_height, ops.Tensor) and offset_height < 0:
```
Using `isinstance(something, ops.Tensor)` here is too clumsy.

Solution 1:
Use some helper function to check whether the value is a "literal" or "symbol".

Solution 2:
Make the framework automatically evaluate all ops that have constant inputs at graph construction time (like C++ constexpr).

martinwicke replied:
Solution 2 is not something that will happen right now (although constant folding does happen before execution)

How about making _is_tensor (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/framework/python/framework/tensor_util.py#L218) public and using that?

gaohuazuo committed:
`resize_image_with_crop_or_pad` minor change

Changed to accept both tensor and integer as `target_height` and `target_width`

gaohuazuo committed:
bugfix

gaohuazuo committed:
Test case update for `resize_image_with_crop_or_pad`

gaohuazuo committed:
bugfix

gaohuazuo committed:
Updated test cases for `crop_to_bounding_box` and `pad_to_bounding_box`

vrv commented:
@gaohuazuo you'll have to ping us when this is ready for review

gaohuazuo committed:
Made `is_tensor` public

gaohuazuo committed:
Added doc for `is_tensor`

gaohuazuo committed:
Added a batch mode for `is_tensor`

gaohuazuo committed:
Refactor: use `is_tensor`

gaohuazuo committed:
Reverted 058f024 and minor change

gaohuazuo committed:
Typo fixes and stylistic improvements

gaohuazuo commented:
@martinwicke Ready for review.

martinwicke commented on tensorflow/contrib/framework/python/framework/tensor_util.py:
```diff
@@ -215,8 +216,19 @@ def with_same_shape(expected_tensor, tensor):
-  return isinstance(t, (ops.Tensor, ops.SparseTensor, variables.Variable))
+def is_tensor(x):
+  """Check for tensor types.
+  If the object represents a tensor, returns True. Equivalent to
```
`True`

martinwicke commented on tensorflow/contrib/framework/python/framework/tensor_util.py:
```diff
@@ -215,8 +216,19 @@ def with_same_shape(expected_tensor, tensor):
+    x: An python object to check.
+
+  Returns:
+    True if `x` represents a tensor, False if not.
```
`True`, `False`

martinwicke commented on tensorflow/python/ops/image_ops.py:
```diff
@@ -194,12 +195,18 @@ def _ImageDimensions(images):
 
   Returns:
-    list of integers [batch, height, width, channels]
+    list of integers [batch, height, width, channels], when static shape is
```
`[batch, height, width, channels]`

martinwicke commented on tensorflow/python/ops/image_ops.py:
```diff
@@ -194,12 +195,18 @@ def _ImageDimensions(images):
-    list of integers [batch, height, width, channels]
+    list of integers [batch, height, width, channels], when static shape is
+    fully defined.
+    list of integer scalar tensors [batch, height, width, channels], when
```
here too

martinwicke commented on tensorflow/python/ops/image_ops.py:
```diff
@@ -427,36 +434,47 @@ def pad_to_bounding_box(image, offset_height, offset_width, target_height,
+  if not is_tensor(offset_width) and offset_width < 0:
+    raise ValueError('offset_width must be >= 0')
+
+  if (all(not is_tensor(i) for i in [target_width, width]) and
```
for sets of 2, is_tensor(target_width) and is_tensor(width) is much easier to read

martinwicke commented on tensorflow/python/ops/image_ops.py:
```diff
@@ -427,36 +434,47 @@ def pad_to_bounding_box(image, offset_height, offset_width, target_height,
+  if (all(not is_tensor(i) for i in [target_width, width]) and
+      target_width < width):
+    raise ValueError('target_width must be >= width')
+  if (all(not is_tensor(i) for i in [target_height, height]) and
```
here too

martinwicke commented on tensorflow/python/ops/image_ops_test.py:
```diff
@@ -551,21 +567,31 @@ def testError(self):
 class PadToBoundingBoxTest(test_util.TensorFlowTestCase):
```
Can you add a test that checks that it is possible to fail these various checks? I know that should have existed before but now we need it even more.

martinwicke commented:
Thanks! Added bunch of comments. Mainly: Some doc nits, some nit nits, and, could you add another test that assertRaises appropriately if you pass in illegal values, and with also asserts that if you pass in tensors which have illegal values it still fails (although not with ValueError, but probably with OpError).

gaohuazuo committed:
Typo in `is_tensor`

gaohuazuo committed:
Typo fix

gaohuazuo committed:
Refactored a few resizing ops

New function added: `_assert`
Run time checks added: `pad_to_bounding_box` `crop_to_bounding_box` `resize_with_crop_or_pad`

gaohuazuo committed:
Rewrote test cases for a few resizing ops

`PadToBoundingBoxTest` `CropToBoundingBoxTest` `ResizeImageWithCropOrPadTest`

gaohuazuo committed:
doc improvement for `is_tensor`

gaohuazuo commented:
@martinwicke Test cases updated. Also added corresponding run time check for every static check, though not sure if these checks worth the cost of more complexity in the graph.

gaohuazuo commented on tensorflow/python/ops/image_ops.py:
```diff
@@ -212,16 +247,26 @@ def _Check3DImage(image, require_static=True):
 
   Raises:
     ValueError: if image.shape is not a [3] vector.
```
`image.shape`

gaohuazuo commented on tensorflow/python/ops/image_ops.py:
```diff
@@ -212,16 +247,26 @@ def _Check3DImage(image, require_static=True):
     ValueError: if image.shape is not a [3] vector.
+
+  Returns:
+    An empty list, if image has fully defined dimensions. Otherwise, a list
```
`image`

gaohuazuo commented on tensorflow/python/ops/image_ops.py:
```diff
@@ -484,24 +532,38 @@ def crop_to_bounding_box(image, offset_height, offset_width, target_height,
+    array_ops.pack([target_height, target_width, -1]))
+
+  cropped_shape = [None if is_tensor(i) else i
+                  for i in [target_height, target_width, depth]]
```
misalignment

gaohuazuo commented on tensorflow/python/ops/image_ops.py:
```diff
@@ -531,43 +593,74 @@ def resize_image_with_crop_or_pad(image, target_height, target_width):
+
+  image = control_flow_ops.with_dependencies(assert_ops, image)
+  # `crop_to_bounding_box` and `pad_to_bounding_box` have their own checks.
+  # Make sure our checks comes first, so that error messages are clearer.
```
\# Make sure our checks **come** first ...

gaohuazuo commented on tensorflow/python/ops/image_ops.py:
```diff
@@ -531,43 +593,74 @@ def resize_image_with_crop_or_pad(image, target_height, target_width):
+  # Make sure our checks comes first, so that error messages are clearer.
+  if is_tensor(target_height):
+    target_height = control_flow_ops.with_dependencies(
+      assert_ops, target_height)
```
line wrap not necessary

gaohuazuo commented on tensorflow/python/ops/image_ops_test.py:
```diff
@@ -550,45 +655,148 @@ def testError(self):
+                   ([0,  2, 4, 4], 'width must be <= target - offset'))
+
+    for params, err_msg in test_config:
+      self._assertRaises(x, x_shape, *params, err_msg)
```
python 2 compatibility

gaohuazuo committed:
Typo fixes

gaohuazuo committed:
bugfix

martinwicke commented:
Jenkins, test this please.

martinwicke commented:
Jenkins, test this please.

martinwicke commented:
There are some failing tests in this, can you check?

gaohuazuo committed:
Added missing import

gaohuazuo committed:
Resolved python 3.5- compatibility issue

gaohuazuo committed:
Changed `_is_tensor` to `is_tensor`

gaohuazuo commented:
@martinwicke `//tensorflow/python/image_ops_test` and `//tensorflow/contrib/framework:tensor_util_test` fixed. There are other failing tests, such as `//tensorflow/core/platform/cloud:oauth_client_test`, but I believe they are irrelevant to the PR.

martinwicke commented:
Sorry, apparently we accumulated conflicts. Can you rebase?

gaohuazuo committed:
Merge branch 'master' of https://github.com/tensorflow/tensorflow

gaohuazuo commented:
@martinwicke The conflict was caused by `import` statements inserted into the same location. Fixed.

rmlarsen commented:
It looks like there are still conflicts. Can you check?

gaohuazuo committed:
Merge https://github.com/tensorflow/tensorflow

gaohuazuo commented:
Conflicts resolved.

martinwicke commented:
Jenkins, test this please

martinwicke commented:
Excellent! Thank you!
